### ISSUE TITLE
Uyuni Bugreport - Uyuni 2021.12 - system.bootstrap API call broken ?

### ISSUE THREAD

**Issue by phibid** [uyuni-project/uyuni#4737]
I have an issue with 2021.12 trying to register a server using the API.
This command was working in the previous Uyuni versions:

```
$ spacecmd -- api -A "hostname,22,root,root_pass,registration_key,proxy_id,1" system.bootstrap
```

However this is not working anymore and I have this error in the Uyuni log:

`com.redhat.rhn.common.translation.TranslationException: Could not find translator for class java.lang.String to class java.lang.Integer`

In the API doc, I can see that a SessionKey has to be specified ?

Is it something new or am I forgetting something else?

**Comment by mcalmer** [comment]
I think the last parameter is the problem. You give now an integer, but a boolean is required.
We made a change which is now a bit more picky when it comes to types. In the past we were more "guessing" what function we should call.
Good news: I am working on official bootstrap support in spacecmd. So in future there is no need to use "api" subcommand.

**Comment by mcalmer** [comment]
I checked spacecmd code: there is a function `string_to_bool` which match on "true" or "yes". Try this for the last parameter

**Comment by phibid** [comment]
Ah great to hear that we will not have to use the API anymore :-)

No, no luck:

```
$ spacecmd -- api -A "centos7,22,root,r00tm3,1-centos7_ssh,1000010024,true" system.bootstrap
```

Same error message:

`com.redhat.rhn.common.translation.TranslationException: Could not find translator for class java.lang.String to class java.lang.Boolean`

**Comment by mcalmer** [comment]
Ok, seems that the parser do not allow "bool" when you just give a simple string list.
Try to give real JSON. Maybe that could work:

`-A "['centos7',22,'root','r00tm3,1-centos7_ssh',1000010024,true]"`

**Comment by phibid** [comment]
No luck neither :-(

Tried with both 'true' and '1':
```
$ spacecmd -- api -A "['centos7',22,'root','r00tm3','1-centos7_ssh',1000010024,true]" system.bootstrap
INFO: Connected to https://uyuni.cms.colt/rpc/api as admin

$ spacecmd -- api -A "['centos7',22,'root','r00tm3','1-centos7_ssh',1000010024,1]" system.bootstrap
INFO: Connected to https://uyuni.cms.colt/rpc/api as admin
```

Still the same error.

**Comment by phibid** [comment]
Somebody else able to reproduce this bug or is this only me ?

**Comment by mcalmer** [comment]
@phibid you can try out the assigned PR. just copy the `system.py` file on your system with spacecmd at `/usr/lib/python3.6/site-packages/spacecmd/system.py`

**Comment by phibid** [comment]
@mcalmer Thanks ! I understand that your PR add the new `system_bootstrap` option feature in spacecmd but does not resolve the issue I have, correct ?

I have tried with the `system_bootstrap` option, but I can't make it work:

```
$ spacecmd -- system_bootstrap --ssh-password r00tm3 --activation-key 1-centos7_ssh --proxyid 1000010024 --saltssh true
INFO: Connected to https://uyuni/rpc/api as admin
*** Unknown syntax: system_bootstrap --ssh-password r00tm3 --activation-key 1-centos7_ssh --proxyid 1000010024 --saltssh true
```

Is the command correct ? Typo or something ?

**Comment by mcalmer** [comment]
`--saltssh` does not have a value in spacecmd. Just diving this option say "True". Not giving this option means "False".

```
$ spacecmd -- system_bootstrap --ssh-password r00tm3 --activation-key 1-centos7_ssh --proxyid 1000010024 --saltssh
```
And you need also `--hostname <FQDN>` for the target host you want to bootstrap

**Comment by phibid** [comment]
Aaaah, sorry, forget it, it works ! I was executing spacecmd from a remote server ... (Yeah, Mondays are hard ...).

Thanks !

**Comment by phibid** [comment]
The fix has not been implemented in 2022.02 I think ?

After the upgrade, I had the same issue, so I have copied again the file [system.py](https://github.com/uyuni-project/uyuni/blob/4930d506ef3fd807eda5aa9af876b31fd98c6fdc/spacecmd/src/spacecmd/system.py) however it failed to register new servers:

```
# spacecmd -u "admin" -p "xx" -- system_bootstrap --hostname ubuntu2004 --ssh-password r00tm3 --activation-key 1-ubuntu2004_ssh --proxyid 1000010024 --saltssh
INFO: Spacewalk Username: admin
INFO: Connected to https://uyuni/rpc/api as admin
Bootstrapping 'ubuntu2004'. This may take a while.
ERROR: redstone.xmlrpc.XmlRpcFault: unhandled internal exception: com.suse.salt.netapi.exception.SaltException: Response code: 500
```

The complete stack trace:

```
2022-03-01 13:42:04,273 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-3] ERROR com.redhat.rhn.frontend.xmlrpc.BaseHandler - Error calling method:
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.redhat.rhn.frontend.xmlrpc.BaseHandler.invoke(BaseHandler.java:145)
        at redstone.xmlrpc.XmlRpcDispatcher.dispatch(XmlRpcDispatcher.java:123)
        at com.redhat.rhn.frontend.xmlrpc.RhnXmlRpcServer.execute(RhnXmlRpcServer.java:53)
        at com.redhat.rhn.frontend.xmlrpc.XmlRpcServlet.doPost(XmlRpcServlet.java:164)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.opensymphony.module.sitemesh.filter.PageFilter.parsePage(PageFilter.java:142)
        at com.opensymphony.module.sitemesh.filter.PageFilter.doFilter(PageFilter.java:58)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.redhat.rhn.frontend.servlets.LocalizedEnvironmentFilter.doFilter(LocalizedEnvironmentFilter.java:69)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.redhat.rhn.frontend.servlets.EnvironmentFilter.doFilter(EnvironmentFilter.java:99)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.redhat.rhn.frontend.servlets.SessionFilter.doFilter(SessionFilter.java:55)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.redhat.rhn.frontend.servlets.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:97)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.ajp.AjpProcessor.service(AjpProcessor.java:432)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1591)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: Response code: 500

        at com.suse.manager.webui.services.impl.SaltService.callSyncResult(SaltService.java:229)
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:238)
        at com.suse.manager.webui.services.impl.SaltService.getGrains(SaltService.java:448)
        at com.suse.manager.reactor.messaging.RegisterMinionEventMessageAction.registerSSHMinion(RegisterMinionEventMessageAction.java:156)
        at com.suse.manager.webui.controllers.utils.SSHMinionBootstrapper.bootstrapInternal(SSHMinionBootstrapper.java:111)
        at com.suse.manager.webui.controllers.utils.AbstractMinionBootstrapper.bootstrap(AbstractMinionBootstrapper.java:100)
        at com.redhat.rhn.frontend.xmlrpc.system.XmlRpcSystemHelper.lambda$bootstrap$1(XmlRpcSystemHelper.java:133)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.Streams$StreamBuilderImpl.tryAdvance(Streams.java:397)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.findAny(ReferencePipeline.java:548)
        at com.redhat.rhn.frontend.xmlrpc.system.XmlRpcSystemHelper.bootstrap(XmlRpcSystemHelper.java:138)
        at com.redhat.rhn.frontend.xmlrpc.system.SystemHandler.bootstrap(SystemHandler.java:7940)
        ... 48 more
´´´

What I am missing ?

**Comment by phibid** [comment]
Regarding the error 500, I have opened a separate bug, https://github.com/uyuni-project/uyuni/issues/4737, as it not seems to be related with the spacecmd issue.
